lib/bindings/c/src/lib.rs

```diff
@@ -306,33 +315,593 @@ pub extern "C" fn dynamo_kv_event_publish_removed(
-//                 1,  // worker_id
-//                 32, // kv_block_size
-//             )
-//         };
-
-//         assert_eq!(result as u32, DynamoLlmResult::OK as u32);
-
-//         assert!(WK.get().is_some());
-
-//         let shutdown_result = dynamo_llm_shutdown();
-//         assert_eq!(shutdown_result as u32, DynamoLlmResult::OK as u32);
-//     }
+// KV Router configuration structure for C API
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct DynamoKvRouterConfig {
+    pub overlap_score_weight: f64,
+    pub router_temperature: f64,
+    pub use_kv_events: bool,
+    pub router_replica_sync: bool,
+    pub max_num_batched_tokens: u32,
+}
+
+impl Default for DynamoKvRouterConfig {
+    fn default() -> Self {
+        Self {
+            overlap_score_weight: 1.0,
+            router_temperature: 0.0,
+            use_kv_events: true,
+            router_replica_sync: false,
+            max_num_batched_tokens: 8192,
+        }
+    }
+}
+
+// Helper function to convert DynamoKvRouterConfig to KvRouterConfig
+fn convert_to_kv_router_config(config: DynamoKvRouterConfig) -> KvRouterConfig {
+    KvRouterConfig::new(
+        Some(config.overlap_score_weight),
+        Some(config.router_temperature),
+        Some(config.use_kv_events),
+        Some(config.router_replica_sync),
+        Some(config.max_num_batched_tokens),
+        Some(Some(10000)), // router_snapshot_threshold - use default
+        Some(false),       // router_reset_states - use default
+    )
+}
+
+// KV Router per-request override structure for C API
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct DynamoRouterConfigOverride {
+    pub has_overlap_score_weight: bool,
+    pub overlap_score_weight: f64,
+    pub has_router_temperature: bool,
+    pub router_temperature: f64,
+}
+
+// Helper function to convert DynamoRouterConfigOverride to RouterConfigOverride
+fn convert_to_router_config_override(config: DynamoRouterConfigOverride) -> RouterConfigOverride {
+    RouterConfigOverride {
+        overlap_score_weight: if config.has_overlap_score_weight {
+            Some(config.overlap_score_weight)
+        } else {
+            None
+        },
+        router_temperature: if config.has_router_temperature {
+            Some(config.router_temperature)
+        } else {
+            None
+        },
+    }
+}
+
+// KV Router functions
+/// Initialize the KV router with configuration
+/// Block size is auto-discovered from the worker's ModelDeploymentCard
+/// This must be called after dynamo_llm_init() and before any router operations
+#[unsafe(no_mangle)]
+pub extern "C" fn dynamo_kv_router_init_with_config(
+    namespace_c_str: *const c_char,
+    component_c_str: *const c_char,
+    config: *const DynamoKvRouterConfig,
+) -> DynamoLlmResult {
+    let namespace = match unsafe { CStr::from_ptr(namespace_c_str) }.to_str() {
+        Ok(s) => s.to_string(),
+        Err(e) => {
+            eprintln!("Failed to convert namespace C string: {:?}", e);
```

Comment:
Could we still use `tracing`? We'd need to call `tracing_subscriber::fmt::init();` before first call, I don't know if these bindings have a setup/init call. Otherwise we'd need a log function with a `SyncOnce`.

If not, do you want to prefix all the stderr stuff with `dynamo: `, something like that? To make it easier to identify in the callers' logs.